### DISCUSSION TITLE
Eslint in my monorepo does not seem to be working with the new flat config

### DISCUSSION THREAD

**Discussion by acommodari** [eslint/eslint#17232] [Q&A] [answered]
Hey I'm trying to setup a new monorepo project with npm workspaces using the new flat config and I believe I have done it correctly. This is after following the new config docs as well as all the recommended setups for each plugin. (Please let me know if I missed something)

```js
// eslint.config.mjs
import js from '@eslint/js';
import ts from '@typescript-eslint/eslint-plugin';
import parser from '@typescript-eslint/parser';
import prettier from 'eslint-config-prettier';
import jest from 'eslint-plugin-jest';
import sonarjs from 'eslint-plugin-sonarjs';
import globals from 'globals';
import { dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

export default [
	{
		ignores: ['**/tests/coverage'],
	},
	js.configs.recommended,
	ts.configs.recommended,
	ts.configs['recommended-requiring-type-checking'],
	sonarjs.configs.recommended,
	prettier,
	{
		files: ['./services/**/*.ts', './packages/**/*.ts', './libs/**/*.ts', './jest.config.ts'],
		languageOptions: {
			globals: {
				...globals.node,
			},
			ecmaVersion: 2022,
			parser,
			parserOptions: {
				tsconfigRootDir: dirname(fileURLToPath(import.meta.url)),
				project: true,
			},
		},
		linterOptions: {
			noInlineConfig: true,
			reportUnusedDisabledDirectives: true,
		},
		plugins: {
                        '@typescript-eslint': ts,
			sonarjs,
		},
		rules: {},
	},
	{
		files: ['./services/*/tests/**/*.ts', './packages/*/tests/**/*.ts'],
		...jest.configs.recommended,
		...jest.configs.style,
		plugins: {
			jest,
		},
		rules: {},
	},
];

```
However if i run `eslint .` either at the root of my monorepo (project/) or in one of my individual services (project/services/notifications) I get the following error: 

```
Oops! Something went wrong! :(

ESLint: 8.41.0

ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    npm init @eslint/config

ESLint looked for configuration files in /Users/alessandro.commodari/ssense/dm-shipping/services/notifications/tests/coverage/lcov-report and its ancestors. If it found none, it then looked in your home directory.

If you think you already have a configuration file or if you need more help, please stop by the ESLint Discord server: https://eslint.org/chat
```

My goal is to be able to call eslint on the whole monorepo, on a particular service through an npm script `lint: eslint .`, or recursively call each services `lint` script through workspaces.

Any help would be greatly appreciated 🙏

**Comment by acommodari** [comment]
my current dependencies:
`eslint`
`@eslint/js`
`eslint-config-prettier`
`eslint-plugin-jest`
`eslint-plugin-sonarjs`
`globals`
`@typescript-eslint/parser`
`@typescript-eslint/eslint-plugin`
`prettier`

**Reply by acommodari** [reply]
If the issue is with one of the plugins specifically I'm also not opposed to completely removing either `sonarjs` or `jest` but I absolutely need `prettier` and `typescript`. 🙏

**Comment by Rec0iL99** [comment]
Hi @acommodari, ESLint by default searches for `eslint.config.js` and since you are using `eslint.config.mjs`, it can't find a config file and hence complains. 

You must set the `ESLINT_USE_FLAT_CONFIG ` environment variable to `true` and must pass `eslint.config.mjs` to the `-c` option in CLI in order for your case to work.

```bash
ESLINT_USE_FLAT_CONFIG=true npx eslint -c eslint.config.mjs .
```

[docs](https://eslint.org/docs/latest/use/configure/configuration-files-new#configuration-file-resolution)

Thanks for using ESLint.

**Reply by acommodari** [reply]
Thank you so much for the help! Had no idea about the extensions this definitely helps a lot. I tried your command and now and I'm running into a new error 🤔 
```sh
> ESLINT_USE_FLAT_CONFIG=true npx eslint -c eslint.config.mjs . 

Oops! Something went wrong! :(

ESLint: 8.41.0

Error: Unexpected key "extends" found.
    at ObjectSchema.validate (/Users/alessandro.commodari/ssense/dm-shipping/node_modules/@humanwhocodes/object-schema/src/object-schema.js:203:23)
    at /Users/alessandro.commodari/ssense/dm-shipping/node_modules/@humanwhocodes/object-schema/src/object-schema.js:171:18
    at Array.reduce (<anonymous>)
    at ObjectSchema.merge (/Users/alessandro.commodari/ssense/dm-shipping/node_modules/@humanwhocodes/object-schema/src/object-schema.js:169:24)
    at /Users/alessandro.commodari/ssense/dm-shipping/node_modules/@humanwhocodes/config-array/api.js:884:42
    at Array.reduce (<anonymous>)
    at FlatConfigArray.getConfig (/Users/alessandro.commodari/ssense/dm-shipping/node_modules/@humanwhocodes/config-array/api.js:883:39)
    at FlatConfigArray.isFileIgnored (/Users/alessandro.commodari/ssense/dm-shipping/node_modules/@humanwhocodes/config-array/api.js:911:15)
    at /Users/alessandro.commodari/ssense/dm-shipping/node_modules/eslint/lib/eslint/eslint-helpers.js:312:49
    at Array.reduce (<anonymous>)
```

and also I notice that if I try to run your command within a single service of my monorepo say `/services/notifications` it seems like eslint can't find the `eslint.config.mjs` file at the root of the monorepo. I need to explicitly tell it `../../eslint.config.mjs` is this expected behaviour? if I were to change the file to .js extension would it be able to crawl up to the root and find it automatically?

**Reply by acommodari** [reply]
I'm assuming this latest error is some configuration issue on my part and not an issue with `@humanwhocodes/*` modules but I'm not sure what is wrong here as I'm not directly using any `extends`.

**Reply by Rec0iL99** [reply]
That's strange since your config file has no `extends` key.  Can you please double-check? Thanks.

ESLint only searches for `eslint.config.js`. It first searches the CWD, if it can't find one it moves to the next parent directory and continues till it finds a `eslint.config.js` or reaches the root directory. Setting the `ESLINT_USE_FLAT_CONFIG ` env variable stops this search and now you would have to manually tell ESLint where your config file exists, so yes it's working as expected.

**Reply by acommodari** [reply]
Thanks for the info I've decided to go with .js since it fixes a lot of issues
```js
// eslint.config.js
const js = require('@eslint/js');
const ts = require('@typescript-eslint/eslint-plugin');
const parser = require('@typescript-eslint/parser');
const prettier = require('eslint-config-prettier');
const jest = require('eslint-plugin-jest');
const sonarjs = require('eslint-plugin-sonarjs');
const globals = require('globals');

module.exports = [
	{
		ignores: ['**/tests/coverage'],
	},
	js.configs.recommended,
	ts.configs.recommended,
	ts.configs['recommended-requiring-type-checking'],
	sonarjs.configs.recommended,
	prettier,
	{
		files: ['./services/**/*.ts', './packages/**/*.ts', './libs/**/*.ts', './jest.config.ts'],
		languageOptions: {
			globals: {
				...globals.node,
			},
			ecmaVersion: 2022,
			parser,
			parserOptions: {
				tsconfigRootDir: __dirname,
				project: true,
			},
		},
		linterOptions: {
			noInlineConfig: true,
			reportUnusedDisabledDirectives: true,
		},
		plugins: {
			'@typescript-eslint': ts,
			sonarjs,
		},
		rules: {},
	},
	{
		files: ['./services/*/tests/**/*.ts', './packages/*/tests/**/*.ts'],
		...jest.configs.recommended,
		...jest.configs.style,
		plugins: {
			jest,
		},
		rules: {},
	},
];
```
However I'm still getting the extends issue... 🤔 I tried deleting both `package-lock.json` and `node_modules` and reinstalling deps in case there was some funny business happening but that didn't fix it unfortunately.
A quick search with vscode of my whole repo did not come up with any hidden eslint config files that I might have forgotten about.
Maybe this is an issue with a particular plugin or config that I'm using? Or I'm not importing one of them correctly?

**Reply by Rec0iL99** [reply]
> Maybe this is an issue with a particular plugin or config that I'm using? Or I'm not importing one of them correctly?

Could be. You could try commenting out configs one by one and find out which one could be causing the issue :) If that doesn't work please downgrade the ESLint version and see if that helps.

**Reply by acommodari** [reply]
so it seems to only start working after I remove everything except `@eslint/js` recommended configs.

Which is obviously a big issue for me. I doubt every single one of these plugins/configs have an issue so it's definitely me not doing something correctly... but what? Pretty sure the way I was adding these predefined rules to the config file was the same way as shown in the docs. 😵‍💫

```js
const js = require('@eslint/js');
const ts = require('@typescript-eslint/eslint-plugin');
const parser = require('@typescript-eslint/parser');
const prettier = require('eslint-config-prettier');
const jest = require('eslint-plugin-jest');
const sonarjs = require('eslint-plugin-sonarjs');
const globals = require('globals');

module.exports = [
	{
		ignores: ['**/tests/coverage', '**/.serverless', '**/.esbuild'],
	},
	js.configs.recommended,
	// ts.configs.recommended,
	// ts.configs['recommended-requiring-type-checking'],
	// sonarjs.configs.recommended,
	prettier,
	{
		files: ['./services/**/*.ts', './packages/**/*.ts', './libs/**/*.ts', './jest.config.ts', 'eslint.config.js'],
		languageOptions: {
			globals: {
				...globals.node,
			},
			ecmaVersion: 2022,
			parser,
			parserOptions: {
				tsconfigRootDir: __dirname,
				project: true,
			},
		},
		linterOptions: {
			noInlineConfig: true,
			reportUnusedDisableDirectives: true,
		},
		plugins: {
			'@typescript-eslint': ts,
			sonarjs,
		},
		rules: {},
	},
	{
		files: ['./services/*/tests/**/*.ts', './packages/*/tests/**/*.ts'],
		// ...jest.configs.recommended,
		// ...jest.configs.style,
		plugins: {
			jest,
		},
		rules: {},
	},
];
```

**Reply by acommodari** [reply]
actually... digging into `@typescript-eslint/eslint-plugin` I do see that it uses `extends` 
https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/recommended.ts
so maybe this is an issue with all the plugins...

**Reply by Rec0iL99** [reply, marked as answer]
Ah, I could reproduce the error you were facing. The culprit is `typescript-eslint`. They haven't still ported over to flat config ([refer](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/recommended.ts)). There's a PR open https://github.com/typescript-eslint/typescript-eslint/pull/6836 for moving to flat config.

I think you should also face an issue with `eslint-plugin-sonarjs`.

If you just want the recommended rules you can do this:

```js
rules: {
    ...ts.configs.recommended.rules,
    ...ts.configs['recommended-requiring-type-checking'].rules,
    ...sonarjs.configs.recommended.rules,
}
```

Hope that helps.

Flat config is new and will take time to be adopted by the community. You can track the implementation here https://github.com/eslint/eslint/issues/13481. Thanks.

**Reply by Rec0iL99** [reply]
> actually... digging into `@typescript-eslint/eslint-plugin` I do see that it uses `extends` https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/recommended.ts so maybe this is an issue with all the plugins...

Oh, you found it. Nice :)
